File: samples/dnn/mask_rcnn.py
'''
Mask R-CNN
This is an example of using Mask R-CNN for object detection and instance segmentation.

The model is the ONNX Model Zoo build of Mask R-CNN, converted from
facebookresearch/maskrcnn-benchmark. Download it with:

    python download_models.py mask_rcnn

It takes a BGR image of shape (3, height, width) with no batch dimension, resized so
the short side is 800 and zero-padded to a multiple of 32, and returns four tensors:
boxes (nbox, 4), labels (nbox), scores (nbox) and masks (nbox, 1, 28, 28).
'''
import cv2 as cv
import argparse
import numpy as np
import math
import sys
import os

from common import *

def help():
    print(
        '''
        Firstly, download the model using `download_models.py mask_rcnn`. Set the environment
        variable OPENCV_DOWNLOAD_CACHE_DIR to specify where models should be downloaded. Also,
        point OPENCV_SAMPLES_DATA_PATH to opencv/samples/data.

        To run:
            python mask_rcnn.py --input=path/to/your/input/image/or/video (don't pass --input to use device camera)
        '''
    )

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument('--zoo', default=os.path.join(os.path.dirname(os.path.abspath(__file__)), 'models.yml'),
                    help='An optional path to file with preprocessing parameters.')
parser.add_argument('--input', help='Path to input image or video file. Skip this argument to capture frames from a camera.')
parser.add_argument('--thr', type=float, default=0.5, help='Confidence threshold')
parser.add_argument('--colors', help='Optional path to a text file with colors for an every class. '
                                     'An every color is represented with three values from 0 to 255 in BGR channels order.')
args, _ = parser.parse_known_args()
add_preproc_args(args.zoo, parser, 'mask_rcnn')
parser = argparse.ArgumentParser(parents=[parser],
                                 description='Use this script to run Mask-RCNN object detection '
                                             'and instance segmentation.',
                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)
args = parser.parse_args()

if hasattr(args, 'help'):
    help()
    exit(1)

args.model = findModel(args.model, args.sha1)
if args.labels is not None:
    args.labels = findFile(args.labels)

np.random.seed(324)

# Load names of classes. Label 0 is background, so the model's labels are 1-based.
classes = None
if args.labels:
    with open(args.labels, 'rt') as f:
        classes = f.read().rstrip('\n').split('\n')

colors = None
if args.colors:
    with open(args.colors, 'rt') as f:
        colors = [np.array(color.split(' '), np.uint8) for color in f.read().rstrip('\n').split('\n')]

legend = None
def showLegend(classes):
    global legend
    if not classes is None and legend is None:
        blockHeight = 30
        assert(len(classes) == len(colors))

        legend = np.zeros((blockHeight * len(colors), 200, 3), np.uint8)
        for i in range(len(classes)):
            block = legend[i * blockHeight:(i + 1) * blockHeight]
            block[:,:] = colors[i]
            cv.putText(block, classes[i], (0, blockHeight//2), cv.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255))

        cv.namedWindow('Legend', cv.WINDOW_NORMAL)
        cv.imshow('Legend', legend)
        classes = None


def drawBox(frame, classId, conf, left, top, right, bottom):
    cv.rectangle(frame, (left, top), (right, bottom), (0, 255, 0))

    label = '%.2f' % conf
    if classes and 0 <= classId < len(classes):
        label = '%s: %s' % (classes[classId], label)

    labelSize, baseLine = cv.getTextSize(label, cv.FONT_HERSHEY_SIMPLEX, 0.5, 1)
    top = max(top, labelSize[1])
    cv.rectangle(frame, (left, top - labelSize[1]), (left + labelSize[0], top + baseLine), (255, 255, 255), cv.FILLED)
    cv.putText(frame, label, (left, top), cv.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 0))


def preprocess(frame, mean):
    ratio = 800.0 / min(frame.shape[0], frame.shape[1])
    resized = cv.resize(frame, None, fx=ratio, fy=ratio, interpolation=cv.INTER_LINEAR)

    blob = resized.astype(np.float32) - np.array(mean, np.float32)
    blob = blob.transpose(2, 0, 1)

    paddedH = int(math.ceil(blob.shape[1] / 32) * 32)
    paddedW = int(math.ceil(blob.shape[2] / 32) * 32)
    padded = np.zeros((3, paddedH, paddedW), np.float32)
    padded[:, :blob.shape[1], :blob.shape[2]] = blob
    return padded, ratio


net = cv.dnn.readNetFromONNX(args.model)
net.setPreferableBackend(cv.dnn.DNN_BACKEND_OPENCV)
outNames = net.getUnconnectedOutLayersNames()

winName = 'Mask-RCNN in OpenCV'
cv.namedWindow(winName, cv.WINDOW_NORMAL)

cap = cv.VideoCapture(cv.samples.findFileOrKeep(args.input) if args.input else 0)
while cv.waitKey(1) < 0:
    hasFrame, frame = cap.read()
    if not hasFrame:
        cv.waitKey()
        break

    frameH, frameW = frame.shape[0], frame.shape[1]

    blob, ratio = preprocess(frame, args.mean)
    net.setInput(blob)

    t0 = cv.getTickCount()
    boxes, labels, scores, masks = net.forward(outNames)
    t = (cv.getTickCount() - t0) / cv.getTickFrequency()

    boxes = boxes.reshape(-1, 4) / ratio
    labels = labels.reshape(-1)
    scores = scores.reshape(-1)

    if not colors:
        numClasses = len(classes) if classes else 80
        colors = [np.array([0, 0, 0], np.uint8)]
        for i in range(1, numClasses + 1):
            colors.append((colors[i - 1] + np.random.randint(0, 256, [3], np.uint8)) / 2)
        del colors[0]

    boxesToDraw = []
    for i in range(len(scores)):
        if scores[i] <= args.thr:
            continue

        # Model labels are 1-based; index 0 is background.
        classId = int(labels[i]) - 1
        left, top, right, bottom = [int(v) for v in boxes[i]]

        left = max(0, min(left, frameW - 1))
        top = max(0, min(top, frameH - 1))
        right = max(0, min(right, frameW - 1))
        bottom = max(0, min(bottom, frameH - 1))
        if right <= left or bottom <= top:
            continue

        boxesToDraw.append([frame, classId, scores[i], left, top, right, bottom])

        classMask = cv.resize(masks[i][0], (right - left + 1, bottom - top + 1))
        mask = (classMask > 0.5)

        roi = frame[top:bottom+1, left:right+1][mask]
        frame[top:bottom+1, left:right+1][mask] = (0.7 * colors[classId % len(colors)] + 0.3 * roi).astype(np.uint8)

    for box in boxesToDraw:
        drawBox(*box)

    label = 'Inference time: %.2f ms' % (t * 1000.0)
    cv.putText(frame, label, (0, 15), cv.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0))

    showLegend(classes)

    cv.imshow(winName, frame)

File: samples/dnn/mask_rcnn.cpp
// Deep learning based Object Detection and Instance Segmentation using Mask R-CNN
// Downloads:
// http://download.tensorflow.org/models/object_detection/mask_rcnn_inception_v2_coco_2018_01_28.tar.gz
// https://raw.githubusercontent.com/opencv/opencv_extra/master/testdata/dnn/mask_rcnn_inception_v2_coco_2018_01_28.pbtxt
// https://raw.githubusercontent.com/spmallick/learnopencv/master/Mask-RCNN/mscoco_labels.names
// It is based on the OpenCV project. It is subject to the license terms in the LICENSE file found in this distribution and at http://opencv.org/license.html
//

#include <fstream>
#include <sstream>
#include <iostream>
#include <string>

#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

using namespace cv;
using namespace dnn;
using namespace std;

string keys =
    "{ help        | | Print help message. }"
    "{ image       |<none>| Path to input image file.  }"
    "{ input i     | 0 | Path to input image file or video or camera id.  }"
    "{ classes     |mscoco_labels.names| Path to a text file with names of classes. }"
    "{ model  m    |mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb | The pre-trained weights.  }"
    "{ height h    |800| Preprocess input image by resizing to a specific height }"
    "{ width w     |800| Preprocess input image by resizing to a specific columns }"
    "{ config c    |mask_rcnn_inception_v2_coco_2018_01_28.pbtxt |The text graph file that has been tuned by the OpenCV’s DNN support group  }"
    "{ cthr        | .5 | Confidence threshold. }"
    "{ mthr        | .4 | Mask threshold. }";

void static drawBox(Mat& frame, int classId, float conf, Rect box, Mat& objectMask, float maskThreshold, vector<Scalar>& colors, vector<string>& classes);

void static postprocess(Mat& frame, const vector<Mat>& outs, float confThreshold, float maskThreshold, vector<Scalar>& colors, vector<string>& classes);

int main(int argc, char** argv)
{
    CommandLineParser parser(argc, argv, keys);
    parser.about("This sample demonstrates instance segmentation network called Mask-RCNN.");
    if (parser.has("help"))
    {
        parser.printMessage();
        return 0;
    }

    float confThreshold;
    float maskThreshold;
    vector<string> classes;
    vector<Scalar> colors;
    int height;
    int width;

    confThreshold = parser.get<float>("cthr");
    maskThreshold = parser.get<float>("mthr");

    string input = parser.get<string>("input");
    string textGraph = parser.get<string>("config");
    string modelWeights = parser.get<string>("model");
    string classesFile = parser.get<string>("classes");

    height = parser.get<int>("height");
    width = parser.get<int>("width");


    ifstream ifs(classesFile.c_str());
    string line;
    while (getline(ifs, line)) classes.push_back(line);

    for(int _c =0; _c<=10; _c++)
    {
        colors.push_back( Scalar(rand() % 255,rand() % 255,rand() % 255 ,255.0));
    }

    Net net = readNet(modelWeights, textGraph);

    string str;
    VideoCapture cap;
    Mat frame, blob;

    try {
        if (parser.has("image"))
        {
            str = parser.get<String>("image");
            frame = imread(str);
        }
        else if (input.find_first_not_of("0123456789") == std::string::npos)
            cap.open(atoi(input.c_str()));
        else
            cap.open(input);
    }
    catch(...) {
        cout << "Could not open the input image/video stream" << endl;
        return 0;
    }

    while (waitKey(1) < 0)
    {
        if (!parser.has("image"))
            cap >> frame;

        if (frame.empty()) {
            cout << "Done processing !!!" << endl;
            waitKey(0);
            break;
        }

        blobFromImage(frame, blob, 1.0, Size(width, height), Scalar(), true, false);

        net.setInput(blob);

        std::vector<String> outNames(2);
        outNames[0] = "detection_out_final";
        outNames[1] = "detection_masks";
        vector<Mat> outs;
        net.forward(outs, outNames);

        postprocess(frame, outs,maskThreshold,confThreshold,colors,classes);

        vector<double> layersTimes;
        double freq = getTickFrequency() / 1000;
        double t = net.getPerfProfile(layersTimes) / freq;
        string label = format("OpenCV Mask R-CNN : %0.0f ms", t);
        putText(frame, label, Point(0, 15), FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0, 0, 0));

        imshow("Mask R-CNN sample", frame);
        if (parser.has("image"))
            frame.release();
    }

    cap.release();

    return 0;
}

void postprocess(Mat& frame, const vector<Mat>& outs, float confThreshold, float maskThreshold, vector<Scalar>& colors, vector<string>& classes )
{
    Mat outDetections = outs[0];
    Mat outMasks = outs[1];

    // Output size of masks is NxCxHxW where
    // N - number of detected boxes
    // C - number of classes (excluding background)
    // HxW - segmentation shape
    const int numDetections = outDetections.size[2];

    outDetections = outDetections.reshape(1,(int)(outDetections.total() / 7));
    for ( int i = 0; i < numDetections; ++i)
    {
        float score = outDetections.at<float>(i, 2);
        if (score > confThreshold)
        {
            int classId = static_cast<int>(outDetections.at<float>(i, 1));
            int left = static_cast<int>(frame.cols * outDetections.at<float>(i, 3));
            int top = static_cast<int>(frame.rows * outDetections.at<float>(i, 4));
            int right = static_cast<int>(frame.cols * outDetections.at<float>(i, 5));
            int bottom = static_cast<int>(frame.rows * outDetections.at<float>(i, 6));

            Rect box(Point(left, top), Point(right, bottom));
            Rect imageBox(0, 0, frame.cols - 1, frame.rows - 1);
            box &= imageBox;

            Mat objectMask(outMasks.size[2], outMasks.size[3],CV_32FC1, outMasks.ptr<float>(i,classId));

            drawBox(frame, classId, score, box, objectMask, maskThreshold, colors, classes);
        }
    }
}

void drawBox(Mat& frame, int classId, float conf, Rect box, Mat& objectMask, float maskThreshold, vector<Scalar>& colors, vector<string>& classes)
{
    rectangle(frame, Point(box.x, box.y), Point(box.x + box.width, box.y + box.height), Scalar(255, 178, 50), 3);
    string label = format("%.2f", conf);
    if (!classes.empty())
    {
        CV_Assert(classId < (int)classes.size());
        label = classes[classId] + ":" + label;
    }

    int baseLine;
    Size labelSize = getTextSize(label, FONT_HERSHEY_SIMPLEX, 0.5, 1, &baseLine);
    box.y = max(box.y, labelSize.height);
    rectangle(frame, Point(box.x, box.y - static_cast<int>(round(1.5*labelSize.height))), Point(box.x + static_cast<int>(round(1.5*labelSize.width)), box.y + baseLine), Scalar(255, 255, 255), FILLED);
    putText(frame, label, Point(box.x, box.y), FONT_HERSHEY_SIMPLEX, 0.75, Scalar(0,0,0),1);

    Scalar color = colors[classId%colors.size()];

    resize(objectMask, objectMask, Size(box.width, box.height));
    Mat mask = (objectMask > maskThreshold);
    Mat roi = frame(box);
    roi = roi * 0.7 + color * 0.3;
    roi.convertTo(roi, CV_8UC3);

    Mat hierarchy;
    vector<vector<Point>> contours;
    findContours(mask, contours, RETR_TREE, CHAIN_APPROX_SIMPLE);
    polylines(roi, contours, true, color, 2, LINE_8);
}
